# T05

Gonzalo Cano Padilla

## 1.1 Regresión con MPG como salida

In [42]:
import pandas as pd
datos = pd.read_excel('Motor Trend Car Road Tests.xlsx')
datos = datos.drop(columns = 'model')
datos

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0   21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1   21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2   22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3   21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4   18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2
5   18.1    6  225.0  105  2.76  3.460  20.22   1   0     3     1
6   14.3    8  360.0  245  3.21  3.570  15.84   0   0     3     4
7   24.4    4  146.7   62  3.69  3.190  20.00   1   0     4     2
8   22.8    4  140.8   95  3.92  3.150  22.90   1   0     4     2
9   19.2    6  167.6  123  3.92  3.440  18.30   1   0     4     4
10  17.8    6  167.6  123  3.92  3.440  18.90   1   0     4     4
11  16.4    8  275.8  180  3.07  4.070  17.40   0   0     3     3
12  17.3    8  275.8  180  3.07  3.730  17.60   0   0     3     3
13  15.2    8  275.8  180  3.07  3.780  18.00   0   0     3     3
14  10.4    8  472.0  205  2.93  5.250  17.98   0   0     3     4
15  10.4    8  460.0  215  3.00  5.424  17.82   0   0     3     4
16  14.7    8  440.0  230  3.23  5.345  17.42   0   0     3     4
17  32.4    4   78.7   66  4.08  2.200  19.47   1   1     4     1
18  30.4    4   75.7   52  4.93  1.615  18.52   1   1     4     2
19  33.9    4   71.1   65  4.22  1.835  19.90   1   1     4     1
20  21.5    4  120.1   97  3.70  2.465  20.01   1   0     3     1
21  15.5    8  318.0  150  2.76  3.520  16.87   0   0     3     2
22  15.2    8  304.0  150  3.15  3.435  17.30   0   0     3     2
23  13.3    8  350.0  245  3.73  3.840  15.41   0   0     3     4
24  19.2    8  400.0  175  3.08  3.845  17.05   0   0     3     2
25  27.3    4   79.0   66  4.08  1.935  18.90   1   1     4     1
26  26.0    4  120.3   91  4.43  2.140  16.70   0   1     5     2
27  30.4    4   95.1  113  3.77  1.513  16.90   1   1     5     2
28  15.8    8  351.0  264  4.22  3.170  14.50   0   1     5     4
29  19.7    6  145.0  175  3.62  2.770  15.50   0   1     5     6
30  15.0    8  301.0  335  3.54  3.570  14.60   0   1     5     8
31  21.4    4  121.0  109  4.11  2.780  18.60   1   1     4     2

In [43]:
# Primera regresión

import statsmodels.api as sm
import numpy as np

y = datos['mpg']
x = datos.drop(columns = 'mpg')
n = len(datos)
unos = np.ones([n,1])

X=np.hstack([unos,x])
ols = sm.OLS(y,X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           3.79e-07
Time:                        16:19:21   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.3034     18.718      0.657      0.518     -26.623      51.229
x1            -0.1114      1.045     -0.107      0.916      -2.285       2.062
x2             0.0133      0.018      0.747      0.463      -0.024       0.050
x3            -0.0215      0.022     -0.987      0.335      -0.067       0.024
x4             0.7871      1.635      0.481      0.635      -2.614       4.188
x5            -3.7153      1.894     -1.961      0.063      -7.655       0.224
x6             0.8210      0.731      1.123      0.274      -0.699       2.341
x7             0.3178      2.105      0.151      0.881      -4.059       4.694
x8             2.5202      2.057      1.225      0.234      -1.757       6.797
x9             0.6554      1.493      0.439      0.665      -2.450       3.761
x10           -0.1994      0.829     -0.241      0.812      -1.923       1.524
==============================================================================
Omnibus:                        1.907   Durbin-Watson:                   1.861
Prob(Omnibus):                  0.385   Jarque-Bera (JB):                1.747
Skew:                           0.521   Prob(JB):                        0.418
Kurtosis:                       2.526   Cond. No.                     1.22e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.22e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Interpretación resultados regresión **

* R^2: podemos ver que tenemos una R^2 de 0.869 la cual nos indica que tiene un buen ajuste pero viendo los valores de los coeficientes podemos ver que la gran mayoria no son estadisticamente significativos, salvo x5 que es el que tiene el p-value más bajo y pudiera ser significativo si bajara ese p-value
* También podemos ver que la mayoria tiene signos negativos o son menores a 1 lo que nos indica que afectan de manera negativa


## Train test split

In [44]:
# Train

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X_train, X_test, y_train, y_test = train_test_split(x, y, train_size=0.4,random_state=137)

# Esclamiento (preguntar al profe)
scaler = StandardScaler().fit(X_train)
X_scaled_train= scaler.transform(X_train)

n=len(X_scaled_train)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_train])
ols = sm.OLS(y_train,X)
results = ols.fit()
results.summary()

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.891
Method:                 Least Squares   F-statistic:                     9.981
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.242
Time:                        16:19:23   Log-Likelihood:                -9.9811
No. Observations:                  12   AIC:                             41.96
Df Residuals:                       1   BIC:                             47.30
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333      0.556     31.361      0.020      10.370      24.497
x1            40.7571     54.010      0.755      0.588    -645.499     727.013
x2            -2.4635      3.021     -0.815      0.565     -40.854      35.927
x3           -10.1581     19.484     -0.521      0.694    -257.727     237.411
x4            14.1135     13.469      1.048      0.485    -157.022     185.249
x5            -0.8005      3.106     -0.258      0.839     -40.266      38.665
x6             1.5727      3.475      0.453      0.729     -42.583      45.729
x7            32.4926     41.262      0.787      0.575    -491.795     556.780
x8            36.1394     42.669      0.847      0.553    -506.022     578.301
x9           -38.3551     43.625     -0.879      0.541    -592.668     515.958
x10           11.9707     17.482      0.685      0.618    -210.157     234.098
==============================================================================
Omnibus:                        0.259   Durbin-Watson:                   2.354
Prob(Omnibus):                  0.879   Jarque-Bera (JB):                0.413
Skew:                           0.212   Prob(JB):                        0.813
Kurtosis:                       2.196   Cond. No.                         382.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Interpretación datos train**

* Con los datos train podemos ver que tenemos un R^2 casi perfecto de 0.99 el cual nos indica que el modelo es casi perfecto o puede estar sobreajustado, pero a pesar de eso si vemos los valores de los coeficientes tenemos algunos valores muy elevados y algunos valores negativos o menores a 0 y sus p-values nos muestran que no son estadisticamente significativos

In [47]:
# Test
scaler = StandardScaler().fit(X_test)
X_scaled_test= scaler.transform(X_test)

n=len(X_scaled_test)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_test])
ols = sm.OLS(y_test,X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.901
Model:                            OLS   Adj. R-squared:                  0.792
Method:                 Least Squares   F-statistic:                     8.231
Date:                Thu, 18 Sep 2025   Prob (F-statistic):            0.00201
Time:                        16:19:49   Log-Likelihood:                -39.494
No. Observations:                  20   AIC:                             101.0
Df Residuals:                       9   BIC:                             111.9
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         21.6850      0.581     37.317      0.000      20.370      23.000
x1             1.6940      2.799      0.605      0.560      -4.639       8.027
x2             2.9928      3.146      0.951      0.366      -4.124      10.109
x3            -1.6543      1.645     -1.005      0.341      -5.376       2.068
x4             0.6034      1.150      0.525      0.613      -1.998       3.205
x5            -4.4365      2.429     -1.826      0.101      -9.932       1.059
x6             1.5627      1.622      0.964      0.360      -2.106       5.231
x7             0.6419      1.275      0.503      0.627      -2.243       3.526
x8             0.8003      1.282      0.624      0.548      -2.099       3.700
x9             2.0424      1.503      1.359      0.207      -1.357       5.442
x10           -0.8071      1.678     -0.481      0.642      -4.602       2.988
==============================================================================
Omnibus:                        7.229   Durbin-Watson:                   2.199
Prob(Omnibus):                  0.027   Jarque-Bera (JB):                4.624
Skew:                           0.963   Prob(JB):                       0.0990
Kurtosis:                       4.356   Cond. No.                         18.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Interpretación datos test**

* La R^2 de los datos test sigue siendo muy alta y nos sigue mostrando que el modelo tiene un buen ajuste pero si vemos los p-values de las variables podemos ver que solo la constante es estadisticamente significativa y los coeficientes no.
* La mayoria de los signos son negativos y los valores menores a 1 lo cual nos muestra que la mayoria de los coeficientes nos afectan de manera negativa al resultado

In [48]:
# Regresión con penalización L2
from sklearn.linear_model import Ridge
from scipy import stats

y = datos['mpg']
x = datos.drop(columns = 'mpg')
n,p = x.shape

L2 = Ridge(alpha=0.01, fit_intercept=True)
L2.fit(x,y)

# RSS
y_hat = L2.predict(x)
RSS_L2 = np.sum((y-y_hat)**2)

#RSE
RSE_L2 = np.sqrt(RSS_L2/(n-p))

# Var_beta
X = np.column_stack([np.ones(n),x]) # Es la x con el intercepto
var_beta = np.linalg.inv(X.T @ X) * RSE_L2**2

std_beta = np.sqrt(var_beta.diagonal())

# Estadisticos t
beta = np.r_[L2.intercept_, L2.coef_]   # vector (p+1,)
t_L2 = beta / std_beta

# P-values
p_L2 = 2 * (1 - stats.t.cdf(np.abs(t_L2), n-p-1))

# R^2
R2 = L2.score(x, y)

print("----Resultados Ridge(L2)----")
print(f"Intercepto: {L2.intercept_}")
print(f"Coefficientes: {L2.coef_}")
print(f"RSS: {RSS_L2}")
print(f"RSE: {RSE_L2}")
print(f"P-Values: {p_L2}")
print(f"R^2: {R2}")


----Resultados Ridge(L2)----
Intercepto: 12.38592271795689
Coefficientes: [-0.11375259  0.0131965  -0.0214229   0.78889495 -3.69840443  0.81545566
  0.31536949  2.50779119  0.65956112 -0.20482935]
RSS: 147.4953648680979
RSE: 2.5892730189633775
P-Values: [0.50561823 0.91234692 0.45782501 0.32526825 0.62661086 0.05880842
 0.2662986  0.87956278 0.22575059 0.65584293 0.80275185]
R^2: 0.8690149342714841


**Interpretación regresión con penalización**

* Podemos ver que en este caso la R^2 bajo pero los valores de los p-values nos siguen mostrando que los datos no son estadisticamente significativos y seguimos teniendo valores negativos y menores a 1 en los coeficientes que no aportan nada positivo a los resultados.

## 1.2 Utilizando QSEC como salida

In [49]:
# Primera regresión
y = datos['qsec']
x = datos.drop(columns = 'qsec')
n = len(datos)
unos = np.ones([n,1])

X=np.hstack([unos,x])
ols = sm.OLS(y,X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           2.44e-07
Time:                        16:20:09   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.7762      3.876      4.586      0.000       9.716      25.837
x1             0.0690      0.061      1.123      0.274      -0.059       0.197
x2            -0.3627      0.293     -1.239      0.229      -0.971       0.246
x3            -0.0075      0.005     -1.505      0.147      -0.018       0.003
x4            -0.0016      0.006     -0.242      0.811      -0.015       0.012
x5            -0.1311      0.476     -0.275      0.786      -1.121       0.859
x6             1.4963      0.500      2.990      0.007       0.456       2.537
x7             0.9700      0.573      1.694      0.105      -0.221       2.161
x8            -0.9012      0.585     -1.540      0.139      -2.118       0.316
x9            -0.2013      0.433     -0.465      0.647      -1.101       0.699
x10           -0.2736      0.233     -1.174      0.254      -0.758       0.211
==============================================================================
Omnibus:                       21.069   Durbin-Watson:                   2.573
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               38.291
Skew:                           1.470   Prob(JB):                     4.84e-09
Kurtosis:                       7.481   Cond. No.                     8.77e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.77e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**Interpretación de los resultados**

* En este caso podemos ver que tambien con la variable QSEC obtenemos una R^2 buena mostrandonos un buen ajuste del modelo y en este caso todos los coeficientes no no interesan y perjudican a los resultados de forma negativa, pero x6 es el unico coeficiente que si es significativo y es valor positivo

In [50]:
# Train
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size=0.4,random_state=137)

# Esclamiento (preguntar al profe)
scaler = StandardScaler().fit(X_train)
X_scaled_train= scaler.transform(X_train)

n=len(X_scaled_train)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_train])
ols = sm.OLS(y_train,X)
results = ols.fit()
results.summary()

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.979
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     4.609
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.349
Time:                        16:20:19   Log-Likelihood:               -0.25847
No. Observations:                  12   AIC:                             22.52
Df Residuals:                       1   BIC:                             27.85
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733      0.247     70.269      0.009      14.232      20.515
x1             1.0235      2.262      0.453      0.729     -27.713      29.760
x2            -0.3081     30.092     -0.010      0.993    -382.663     382.047
x3             0.2789      1.711      0.163      0.897     -21.465      22.023
x4            -1.0127      9.720     -0.104      0.934    -124.519     122.494
x5            -1.0356      8.615     -0.120      0.924    -110.497     108.426
x6             0.5976      1.295      0.461      0.725     -15.861      17.057
x7             0.2129     23.358      0.009      0.994    -296.578     297.004
x8            -0.3877     24.867     -0.016      0.990    -316.349     315.574
x9             0.6755     25.827      0.026      0.983    -327.485     328.836
x10           -0.1562      9.422     -0.017      0.989    -119.876     119.563
==============================================================================
Omnibus:                       12.359   Durbin-Watson:                   2.150
Prob(Omnibus):                  0.002   Jarque-Bera (JB):                7.063
Skew:                           1.532   Prob(JB):                       0.0293
Kurtosis:                       5.178   Cond. No.                         517.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Interpretación datos train**

* Con los datos de entrenamiento podemos ver que la R^2 es bastante alta con un valor de 0.979 osea casi perfecto o nos muestra un sobreajuste y todos los coeficientes son insignificativos y tienen valores negativos o menores a 1 lo cual nos perjudica y en este caso la constante es lo unico estadisticamente significativo

In [51]:
# Test
scaler = StandardScaler().fit(X_test)
X_scaled_test= scaler.transform(X_test)

n=len(X_scaled_test)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_test])
ols = sm.OLS(y_test,X)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.884
Model:                            OLS   Adj. R-squared:                  0.754
Method:                 Least Squares   F-statistic:                     6.833
Date:                Thu, 18 Sep 2025   Prob (F-statistic):            0.00399
Time:                        16:20:30   Log-Likelihood:                -17.875
No. Observations:                  20   AIC:                             57.75
Df Residuals:                       9   BIC:                             68.70
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.1340      0.197     91.983      0.000      17.688      18.580
x1             0.5761      0.598      0.964      0.360      -0.776       1.929
x2            -0.6521      0.944     -0.691      0.507      -2.788       1.484
x3            -1.2537      1.039     -1.207      0.258      -3.603       1.096
x4            -0.3070      0.580     -0.530      0.609      -1.619       1.005
x5             0.0111      0.396      0.028      0.978      -0.885       0.907
x6             1.7616      0.766      2.301      0.047       0.030       3.494
x7             0.2870      0.428      0.670      0.519      -0.681       1.255
x8            -0.6857      0.381     -1.800      0.105      -1.547       0.176
x9            -0.2874      0.551     -0.521      0.615      -1.535       0.960
x10           -0.3200      0.567     -0.565      0.586      -1.602       0.961
==============================================================================
Omnibus:                       17.263   Durbin-Watson:                   1.647
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               24.409
Skew:                           1.368   Prob(JB):                     5.01e-06
Kurtosis:                       7.670   Cond. No.                         19.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**Interpretación de datos test**

* En este caso tenemos valores de los coeficientes muy parecidos a los de train, pero la R^2 tiene un valor menor y se ajusta mejor este modelo a pesar de ser menor

In [52]:
y = datos['qsec']
x = datos.drop(columns = 'qsec')
n,p = x.shape

L2 = Ridge(alpha=0.05, fit_intercept=True)
L2.fit(x,y)

# RSS
y_hat = L2.predict(x)
RSS_L2 = np.sum((y-y_hat)**2)

#RSE
RSE_L2 = np.sqrt(RSS_L2/(n-p))

# Var_beta
X = np.column_stack([np.ones(n),x]) # Es la x con el intercepto
var_beta = np.linalg.inv(X.T @ X) * RSE_L2**2

std_beta = np.sqrt(var_beta.diagonal())

# Estadisticos t
beta = np.r_[L2.intercept_, L2.coef_]   # vector (p+1,)
t_L2 = beta / std_beta

# P-values
p_L2 = 2 * (1 - stats.t.cdf(np.abs(t_L2), n-p-1))

# R^2
R2 = L2.score(x, y)

print("----Resultados Ridge(L2)----")
print(f"Intercepto: {L2.intercept_}")
print(f"Coefficientes: {L2.coef_}")
print(f"RSS: {RSS_L2}")
print(f"RSE: {RSE_L2}")
print(f"P-Values: {p_L2}")
print(f"R^2: {R2}")

----Resultados Ridge(L2)----
Intercepto: 17.934879172645395
Coefficientes: [ 0.06763048 -0.36896485 -0.00734046 -0.00170744 -0.13634652  1.47427056
  0.96112092 -0.89341448 -0.20998468 -0.26800877]
RSS: 12.406188097885202
RSE: 0.7509444991562051
P-Values: [1.12148618e-04 2.72808302e-01 2.10921944e-01 1.46706057e-01
 7.89047386e-01 7.72280765e-01 6.59438085e-03 1.00625673e-01
 1.33133979e-01 6.24662943e-01 2.52562410e-01]
R^2: 0.8746699670830781


**Interpretación con penalización**

* También en este caso tenemos valores muy parecidos que los datos de test y la normal, lo cual nos muestra que podriamos usar cualquiera de esos tres metodos, menos train ya que nos muestra un sobreajuste

## 2.1 Realiza una regresión tomando 'mpg' como salida, con dummies

In [64]:
dummies = pd.get_dummies(datos, columns=['cyl','gear','carb'])
dummies = dummies.astype(float).replace({0: -1, 1: 1})
dummies

mpg   disp     hp  drat     wt   qsec   vs   am  cyl_4  cyl_6  cyl_8  \
0   21.0  160.0  110.0  3.90  2.620  16.46 -1.0  1.0   -1.0    1.0   -1.0   
1   21.0  160.0  110.0  3.90  2.875  17.02 -1.0  1.0   -1.0    1.0   -1.0   
2   22.8  108.0   93.0  3.85  2.320  18.61  1.0  1.0    1.0   -1.0   -1.0   
3   21.4  258.0  110.0  3.08  3.215  19.44  1.0 -1.0   -1.0    1.0   -1.0   
4   18.7  360.0  175.0  3.15  3.440  17.02 -1.0 -1.0   -1.0   -1.0    1.0   
5   18.1  225.0  105.0  2.76  3.460  20.22  1.0 -1.0   -1.0    1.0   -1.0   
6   14.3  360.0  245.0  3.21  3.570  15.84 -1.0 -1.0   -1.0   -1.0    1.0   
7   24.4  146.7   62.0  3.69  3.190  20.00  1.0 -1.0    1.0   -1.0   -1.0   
8   22.8  140.8   95.0  3.92  3.150  22.90  1.0 -1.0    1.0   -1.0   -1.0   
9   19.2  167.6  123.0  3.92  3.440  18.30  1.0 -1.0   -1.0    1.0   -1.0   
10  17.8  167.6  123.0  3.92  3.440  18.90  1.0 -1.0   -1.0    1.0   -1.0   
11  16.4  275.8  180.0  3.07  4.070  17.40 -1.0 -1.0   -1.0   -1.0    1.0   
12  17.3  275.8  180.0  3.07  3.730  17.60 -1.0 -1.0   -1.0   -1.0    1.0   
13  15.2  275.8  180.0  3.07  3.780  18.00 -1.0 -1.0   -1.0   -1.0    1.0   
14  10.4  472.0  205.0  2.93  5.250  17.98 -1.0 -1.0   -1.0   -1.0    1.0   
15  10.4  460.0  215.0  3.00  5.424  17.82 -1.0 -1.0   -1.0   -1.0    1.0   
16  14.7  440.0  230.0  3.23  5.345  17.42 -1.0 -1.0   -1.0   -1.0    1.0   
17  32.4   78.7   66.0  4.08  2.200  19.47  1.0  1.0    1.0   -1.0   -1.0   
18  30.4   75.7   52.0  4.93  1.615  18.52  1.0  1.0    1.0   -1.0   -1.0   
19  33.9   71.1   65.0  4.22  1.835  19.90  1.0  1.0    1.0   -1.0   -1.0   
20  21.5  120.1   97.0  3.70  2.465  20.01  1.0 -1.0    1.0   -1.0   -1.0   
21  15.5  318.0  150.0  2.76  3.520  16.87 -1.0 -1.0   -1.0   -1.0    1.0   
22  15.2  304.0  150.0  3.15  3.435  17.30 -1.0 -1.0   -1.0   -1.0    1.0   
23  13.3  350.0  245.0  3.73  3.840  15.41 -1.0 -1.0   -1.0   -1.0    1.0   
24  19.2  400.0  175.0  3.08  3.845  17.05 -1.0 -1.0   -1.0   -1.0    1.0   
25  27.3   79.0   66.0  4.08  1.935  18.90  1.0  1.0    1.0   -1.0   -1.0   
26  26.0  120.3   91.0  4.43  2.140  16.70 -1.0  1.0    1.0   -1.0   -1.0   
27  30.4   95.1  113.0  3.77  1.513  16.90  1.0  1.0    1.0   -1.0   -1.0   
28  15.8  351.0  264.0  4.22  3.170  14.50 -1.0  1.0   -1.0   -1.0    1.0   
29  19.7  145.0  175.0  3.62  2.770  15.50 -1.0  1.0   -1.0    1.0   -1.0   
30  15.0  301.0  335.0  3.54  3.570  14.60 -1.0  1.0   -1.0   -1.0    1.0   
31  21.4  121.0  109.0  4.11  2.780  18.60  1.0  1.0    1.0   -1.0   -1.0   

    gear_3  gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0     -1.0     1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
1     -1.0     1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
2     -1.0     1.0    -1.0     1.0    -1.0    -1.0    -1.0    -1.0    -1.0  
3      1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0    -1.0    -1.0  
4      1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0    -1.0  
5      1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0    -1.0    -1.0  
6      1.0    -1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
7     -1.0     1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0    -1.0  
8     -1.0     1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0    -1.0  
9     -1.0     1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
10    -1.0     1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
11     1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0  
12     1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0  
13     1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0    -1.0  
14     1.0    -1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
15     1.0    -1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
16     1.0    -1.0    -1.0    -1.0    -1.0    -1.0     1.0    -1.0    -1.0  
17    -1.0     1.0    -1.0     1.0    -1.0    -1.0    -1.0    -1.0    -1.0  


In [74]:
# Primera regresión
y = dummies['mpg']
x = dummies.drop(columns = 'mpg')

n = len(x)

unos = np.ones([n,1])
X = np.hstack([unos, x])
ols = sm.OLS(y,X.astype(float))
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           0.000124
Time:                        18:45:41   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          6.4946      4.406      1.474      0.161      -2.896      15.885
x1             0.0355      0.032      1.114      0.283      -0.032       0.104
x2            -0.0705      0.039     -1.788      0.094      -0.155       0.014
x3             1.1828      2.483      0.476      0.641      -4.111       6.476
x4            -4.5298      2.539     -1.784      0.095      -9.941       0.881
x5             0.3678      0.935      0.393      0.700      -1.626       2.362
x6             0.9654      1.436      0.672      0.512      -2.095       4.025
x7             0.6061      1.607      0.377      0.711      -2.819       4.031
x8            -1.6674      1.740     -0.958      0.353      -5.377       2.042
x9            -2.9917      1.925     -1.554      0.141      -7.096       1.112
x10           -1.8355      2.701     -0.679      0.507      -7.593       3.922
x11           -2.7720      2.078     -1.334      0.202      -7.202       1.658
x12           -2.2148      1.521     -1.456      0.166      -5.456       1.026
x13           -1.5078      1.742     -0.866      0.400      -5.221       2.205
x14           -5.5664      3.739     -1.489      0.157     -13.536       2.404
x15           -6.0560      3.731     -1.623      0.125     -14.008       1.896
x16           -4.0665      3.210     -1.267      0.225     -10.908       2.775
x17           -5.0207      3.157     -1.590      0.133     -11.750       1.709
x18           -3.3276      3.289     -1.012      0.328     -10.338       3.683
x19           -1.9412      3.364     -0.577      0.572      -9.111       5.229
==============================================================================
Omnibus:                        2.468   Durbin-Watson:                   2.105
Prob(Omnibus):                  0.291   Jarque-Bera (JB):                1.772
Skew:                           0.576   Prob(JB):                        0.412
Kurtosis:                       3.009   Cond. No.                     3.03e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.24e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [75]:
# Train
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size=0.4,random_state=137)

# Esclamiento
scaler = StandardScaler().fit(X_train)
X_scaled_train= scaler.transform(X_train)

n=len(X_scaled_train)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_train])
ols = sm.OLS(y_train,X.astype(float))
results = ols.fit()
results.summary()

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1796: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1796: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1718: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/base/model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Thu, 18 Sep 2025   Prob (F-statistic):                nan
Time:                        18:45:45   Log-Likelihood:                 367.88
No. Observations:                  12   AIC:                            -711.8
Df Residuals:                       0   BIC:                            -705.9
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333        inf          0        nan         nan         nan
x1            -2.8807        inf         -0        nan         nan         nan
x2             1.1616        inf          0        nan         nan         nan
x3             3.8237        inf          0        nan         nan         nan
x4             2.0417        inf          0        nan         nan         nan
x5            -1.5490        inf         -0        nan         nan         nan
x6            -0.3475        inf         -0        nan         nan         nan
x7             0.5954        inf          0        nan         nan         nan
x8             1.6043        inf          0        nan         nan         nan
x9             0.0494        inf          0        nan         nan         nan
x10           -0.9860        inf         -0        nan         nan         nan
x11            1.4522        inf          0        nan         nan         nan
x12            0.1433        inf          0        nan         nan         nan
x13           -2.0876        inf         -0        nan         nan         nan
x14            2.2705        inf          0        nan         nan         nan
x15            0.0735        inf          0        nan         nan         nan
x16           -0.1647        inf         -0        nan         nan         nan
x17           -1.3885        inf         -0        nan         nan         nan
x18                 0        nan        nan        nan         nan         nan
x19           -0.4010        inf         -0        nan         nan         nan
==============================================================================
Omnibus:                        1.733   Durbin-Watson:                   2.162
Prob(Omnibus):                  0.420   Jarque-Bera (JB):                0.794
Skew:                           0.626   Prob(JB):                        0.672
Kurtosis:                       2.861   Cond. No.                         36.7
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

**Interpretación**

* En este caso tenemos un modelo perfecto y los p-values nan nos indica que el modelo tiene muchas variables para pocos datos con un posible caso de overfit porque el modelo es muy bueno para lo real.

In [76]:
# Test

# Esclamiento
scaler = StandardScaler().fit(X_test)
X_scaled_test= scaler.transform(X_test)

n=len(X_scaled_test)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_test])
ols = sm.OLS(y_test,X.astype(float))
results = ols.fit()
results.summary()

/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1967: RuntimeWarning: divide by zero encountered in scalar divide
  return np.sqrt(eigvals[0]/eigvals[-1])


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.931
Model:                            OLS   Adj. R-squared:                  0.673
Method:                 Least Squares   F-statistic:                     3.604
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.112
Time:                        18:45:47   Log-Likelihood:                -35.912
No. Observations:                  20   AIC:                             103.8
Df Residuals:                       4   BIC:                             119.8
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         21.6850      0.729     29.758      0.000      19.662      23.708
x1             6.7253      6.458      1.041      0.356     -11.205      24.655
x2            -3.7521      2.904     -1.292      0.266     -11.814       4.310
x3             0.7730      1.920      0.403      0.708      -4.557       6.103
x4            -5.9257      4.244     -1.396      0.235     -17.710       5.858
x5             1.3397      2.315      0.579      0.594      -5.088       7.767
x6             1.4302      1.895      0.755      0.493      -3.832       6.693
x7             0.2902      2.340      0.124      0.907      -6.205       6.786
x8            -0.2839      2.788     -0.102      0.924      -8.024       7.457
x9            -0.7616      1.040     -0.733      0.504      -3.648       2.125
x10            0.9745      3.229      0.302      0.778      -7.992       9.941
x11           -1.4951      1.626     -0.919      0.410      -6.010       3.020
x12            0.3004      1.362      0.221      0.836      -3.482       4.083
x13            1.6327      1.399      1.167      0.308      -2.252       5.517
x14           -0.0493      1.451     -0.034      0.975      -4.079       3.981
x15           -0.8432      1.069     -0.789      0.475      -3.812       2.126
x16            0.8130      1.178      0.690      0.528      -2.458       4.084
x17            0.2316      1.535      0.151      0.887      -4.030       4.494
x18            0.7848      1.630      0.482      0.655      -3.740       5.310
x19                 0          0        nan        nan           0           0
==============================================================================
Omnibus:                       12.794   Durbin-Watson:                   2.135
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               13.060
Skew:                           1.155   Prob(JB):                      0.00146
Kurtosis:                       6.215   Cond. No.                          inf
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

## 2.2 QSEC

In [77]:
# Primera regresión
y = dummies['qsec']
x = dummies.drop(columns = 'qsec')

n = len(x)

unos = np.ones([n,1])
X = np.hstack([unos, x])
ols = sm.OLS(y,X.astype(float))
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           4.35e-05
Time:                        18:45:50   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.1841      1.000      3.184      0.006       1.053       5.316
x1             0.0277      0.071      0.393      0.700      -0.123       0.178
x2             0.0035      0.009      0.389      0.703      -0.016       0.023
x3            -0.0021      0.012     -0.173      0.865      -0.027       0.023
x4             0.1079      0.687      0.157      0.877      -1.356       1.571
x5             0.8105      0.739      1.097      0.290      -0.764       2.385
x6             0.1329      0.399      0.333      0.744      -0.717       0.983
x7            -0.8471      0.386     -2.197      0.044      -1.669      -0.025
x8            -0.3828      0.482     -0.794      0.440      -1.411       0.645
x9            -0.9350      0.516     -1.812      0.090      -2.035       0.165
x10           -1.8663      0.579     -3.224      0.006      -3.100      -0.632
x11           -1.3138      0.499     -2.631      0.019      -2.378      -0.250
x12           -0.6477      0.414     -1.566      0.138      -1.529       0.234
x13           -1.2227      0.375     -3.261      0.005      -2.022      -0.423
x14           -1.6304      1.016     -1.604      0.130      -3.797       0.536
x15           -2.0476      0.977     -2.096      0.053      -4.130       0.035
x16           -1.7451      0.811     -2.153      0.048      -3.473      -0.017
x17           -2.6039      0.653     -3.987      0.001      -3.996      -1.212
x18           -2.4130      0.695     -3.471      0.003      -3.895      -0.931
x19           -2.2966      0.722     -3.183      0.006      -3.835      -0.759
==============================================================================
Omnibus:                       12.619   Durbin-Watson:                   2.588
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.989
Skew:                           0.924   Prob(JB):                     0.000205
Kurtosis:                       6.054   Cond. No.                     3.12e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 3.05e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [78]:
# Train
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size=0.4,random_state=137)

# Esclamiento
scaler = StandardScaler().fit(X_train)
X_scaled_train= scaler.transform(X_train)

n=len(X_scaled_train)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_train])
ols = sm.OLS(y_train,X.astype(float))
results = ols.fit()
results.summary()

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  res = hypotest_fun_out(*samples, **kwds)
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1796: RuntimeWarning: divide by zero encountered in divide
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1796: RuntimeWarning: invalid value encountered in scalar multiply
  return 1 - (np.divide(self.nobs - self.k_constant, self.df_resid)
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1718: RuntimeWarning: divide by zero encountered in scalar divide
  return np.dot(wresid, wresid) / self.df_resid
/opt/anaconda3/lib/python3.12/site-packages/statsmodels/base/model.py:1527: RuntimeWarning: invalid value encountered in multiply
  cov_p = self.normalized_cov_params * scale


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       1.000
Model:                            OLS   Adj. R-squared:                    nan
Method:                 Least Squares   F-statistic:                       nan
Date:                Thu, 18 Sep 2025   Prob (F-statistic):                nan
Time:                        18:45:52   Log-Likelihood:                 362.40
No. Observations:                  12   AIC:                            -700.8
Df Residuals:                       0   BIC:                            -695.0
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733        inf          0        nan         nan         nan
x1            -2.8275        inf         -0        nan         nan         nan
x2             2.1520        inf          0        nan         nan         nan
x3            -0.1677        inf         -0        nan         nan         nan
x4             3.7193        inf          0        nan         nan         nan
x5             0.0248        inf          0        nan         nan         nan
x6             0.0007        inf          0        nan         nan         nan
x7            -0.0380        inf         -0        nan         nan         nan
x8             0.7525        inf          0        nan         nan         nan
x9             0.7933        inf          0        nan         nan         nan
x10           -1.1698        inf         -0        nan         nan         nan
x11            1.1786        inf          0        nan         nan         nan
x12            0.4526        inf          0        nan         nan         nan
x13           -2.0851        inf         -0        nan         nan         nan
x14            1.5119        inf          0        nan         nan         nan
x15           -0.1165        inf         -0        nan         nan         nan
x16            0.5565        inf          0        nan         nan         nan
x17           -1.8342        inf         -0        nan         nan         nan
x18                 0        nan        nan        nan         nan         nan
x19            0.6457        inf          0        nan         nan         nan
==============================================================================
Omnibus:                        0.655   Durbin-Watson:                   2.414
Prob(Omnibus):                  0.721   Jarque-Bera (JB):                0.570
Skew:                          -0.054   Prob(JB):                        0.752
Kurtosis:                       1.937   Cond. No.                         113.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
"""

**Interpretación**

* En este caso tenemos un modelo perfecto y los p-values nan nos indica que el modelo tiene muchas variables para pocos datos con un posible caso de overfit porque el modelo es muy bueno para lo real.

In [79]:
# Test

# Esclamiento
scaler = StandardScaler().fit(X_test)
X_scaled_test= scaler.transform(X_test)

n=len(X_scaled_test)
unos = np.ones([n,1])

X = np.hstack([unos, X_scaled_test])
ols = sm.OLS(y_test,X.astype(float))
results = ols.fit()
results.summary()

/opt/anaconda3/lib/python3.12/site-packages/statsmodels/regression/linear_model.py:1967: RuntimeWarning: divide by zero encountered in scalar divide
  return np.sqrt(eigvals[0]/eigvals[-1])


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.909
Model:                            OLS   Adj. R-squared:                  0.566
Method:                 Least Squares   F-statistic:                     2.650
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.179
Time:                        18:46:21   Log-Likelihood:                -15.460
No. Observations:                  20   AIC:                             62.92
Df Residuals:                       4   BIC:                             78.85
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.1340      0.262     69.190      0.000      17.406      18.862
x1             0.5551      0.959      0.579      0.594      -2.108       3.218
x2            -0.3525      2.613     -0.135      0.899      -7.607       6.902
x3             0.2844      1.235      0.230      0.829      -3.145       3.714
x4             0.0065      0.704      0.009      0.993      -1.949       1.962
x5             1.1943      1.763      0.677      0.535      -3.702       6.090
x6             0.0271      0.728      0.037      0.972      -1.996       2.050
x7            -0.9187      0.707     -1.300      0.264      -2.881       1.044
x8             0.7229      0.937      0.772      0.483      -1.878       3.324
x9             0.1552      0.391      0.397      0.711      -0.929       1.240
x10           -0.9242      1.080     -0.856      0.440      -3.923       2.074
x11           -0.0543      0.643     -0.084      0.937      -1.840       1.731
x12            0.3129      0.467      0.670      0.540      -0.985       1.611
x13           -0.3615      0.554     -0.653      0.550      -1.899       1.176
x14            0.2301      0.509      0.452      0.675      -1.184       1.644
x15            0.1383      0.408      0.339      0.751      -0.993       1.270
x16            0.1039      0.445      0.233      0.827      -1.132       1.340
x17           -0.4043      0.515     -0.784      0.477      -1.835       1.027
x18           -0.1348      0.599     -0.225      0.833      -1.798       1.529
x19                 0          0        nan        nan           0           0
==============================================================================
Omnibus:                        7.417   Durbin-Watson:                   1.968
Prob(Omnibus):                  0.025   Jarque-Bera (JB):                8.491
Skew:                           0.179   Prob(JB):                       0.0143
Kurtosis:                       6.172   Cond. No.                          inf
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is      0. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

## 3.2 Comparar resultados de 1.1 y 2.1

* En ambos casos tenemos valores muy parecidos y cada regresión coincide con lo mismo, también podemos ver como los datos train de ambos nos uestran un modelo con un sobreajuste o un modelo casi perfecto
